### PR TITLE
Token voting

### PR DESCRIPTION
**Pull request by Franpastoragusti**
## Description
Create the initial StoryTelling for the Governance Path and the first challenge the Token Voting Contract
based on the specifications from #2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,7 @@
-# Template For Challenge - ETH Tech Tree
+#  Challenge 1: Token Voting Contract - ETH Tech Tree
```
Comment (escottalexander) [inline README.md]: Remove "Challenge 1:". This may or may not be the first challenge in this track.

File: README.md [hunk 2]
```diff
@@ -1,4 +1,7 @@
...
+In a dystopian future where megacorporations have seized control over all aspects of life, a brave group of technologists and activists form an underground movement known as ***The Decentralized Resistance***. Their mission is to create a new society governed by the people, free from the tyranny of corporate overlords. They believe that blockchain technology holds the key to building a fair and transparent governance system. As a key developer in The Decentralized Resistance, you are tasked with creating the smart contracts that will enable this new society to thrive.
```
Comment (escottalexander) [inline README.md:2]: 🤣🤣🤣 This is great!!! Love this touch. We need to add this to every challenge.

File: packages/foundry/contracts/Voting.sol [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        } else {
+            votesAgainst += voterWeight;
+        }
+        hasVoted[msg.sender] = true;
+        emit VoteCasted(msg.sender, support, voterWeight);
+    }
+
+    /**
+     * @dev Function to get the result of the vote
+     * @return The result of the vote as a string ("Proposal Approved" or "Proposal Rejected")
```
Comment (escottalexander) [inline packages/foundry/contracts/Voting.sol]: This isn't what it actually returns.

File: packages/foundry/contracts/Voting.sol [hunk 4]
```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
```
Comment (escottalexander) [inline packages/foundry/contracts/Voting.sol:3]: Add this here for ease of debugging: `import { console2 } from "forge-std/console2.sol";`

It was just merged into the main branch. Nice catch on the README edits that were just merged.

File: packages/foundry/script/Deploy.s.sol
Comment (escottalexander) [inline packages/foundry/script/Deploy.s.sol:1]: We actually need this file to enable the challenger to deploy the contract. You will need to restore the original and then change the contract name, deploy the erc20 and provide params for deployment. 

When you are done, you should be able to run the typical scaffold-eth `yarn chain`, `yarn deploy`, `yarn start` and see the debug menu without any errors when you go to localhost:3000.

File: packages/foundry/test/Voting.sol [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+contract VotingTest is Test {
+    MyToken public token;
+    Voting public voting;
+    address public userOne = address(0x123);
+    address public userTwo = address(0x456);
+
+    function setUp() public {
+        token = new MyToken(1000000 * 10**18); // 1,000,000 tokens
+        voting = new Voting(address(token), 86400); // 1 day voting period
```
Comment (escottalexander) [inline packages/foundry/test/Voting.sol:17]: Make 86400 a state variable at the top of the test and replace everywhere you you use it with the variable.

File: packages/foundry/contracts/Voting.sol [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * @dev Function to get the result of the vote
+     * @return The result of the vote as a string ("Proposal Approved" or "Proposal Rejected")
+     * Requirements:
+     * - Ensure the voting period has ended
+     * - Determine the result based on the majority vote
+     */
+    function getResult() public view returns (bool) {
+        require(block.timestamp >= votingDeadline, "Voting is still ongoing");
+        return votesFor > votesAgainst;
```
Comment (escottalexander) [inline packages/foundry/contracts/Voting.sol:110]: We should specify in the requirements that in the event of a tie the proposal is rejected. I can see a challenger spotting this edge case but not knowing how to handle it without instruction. 

Let's add a test for it too.

File: packages/foundry/test/Voting.sol [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        voting.getResult();
+    }
+    function testVoteAfterDeadline() public {
+        // Try to vote after the voting deadline
+        vm.warp(block.timestamp + 86400 + 1);
+        vm.prank(userOne);
+        vm.expectRevert();
+        voting.vote(true);
+    }
+
```
Comment (escottalexander) [inline packages/foundry/test/Voting.sol:150]: Reminder to add a test to make sure the proposal is rejected when a tie occurs.

File: packages/foundry/test/Voting.sol [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        voting.vote(false);
+
+        vm.prank(userTwo);
+        voting.vote(true);
+
+        vm.warp(block.timestamp + 86400 + 1);
+
+        bool result = voting.getResult();
+        assertEq(result, true);
+    }
```
Comment (escottalexander) [inline packages/foundry/test/Voting.sol:106]: Inconsistent spacing here compared to the rest.

File: packages/foundry/test/Voting.sol [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        // Try to get the result before the voting deadline
+        vm.prank(userOne);
+        voting.vote(true);
+
+        vm.prank(userTwo);
+        voting.vote(false);
+
+        vm.expectRevert();
+        voting.getResult();
+    }
```
Comment (escottalexander) [inline packages/foundry/test/Voting.sol:115]: Inconsistent spacing here too.

File: packages/foundry/contracts/MyToken.sol [hunk 10]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+// Dont touch this contract
+contract MyToken is ERC20 {
+    constructor(uint256 initialSupply) ERC20("MyToken", "MTK") {
```
Comment (escottalexander) [inline packages/foundry/contracts/MyToken.sol]: Since you are being so masterfully playful with the story line I would go ahead and name the token something relevant. "The Decentralized Resistance Token", "RES", "DRT", "✊" Use your imagination. Lol

If you do, make sure you name the contract something to match the token name instead of generic MyToken.

File: packages/foundry/contracts/MyToken.sol [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+        _mint(msg.sender, initialSupply);
+    }
```
Comment (escottalexander) [inline packages/foundry/contracts/MyToken.sol]: I am thinking it might be nice to extend this challenge a little and make it so that if someone transfers their token then their votes are removed.

You would need to provide instruction for the challenger but what I would do is add a override for the transfer function (whichever one is used by both transfer and transferFrom) so that it checks voting balances and undoes the voting action if a address with votes transfers the token. May seem like overkill but right now there is nothing to keep a single person from moving their tokens from address to address and voting each time to amass a vote way bigger than their actual token holdings. Seems like an important edge case and hopefully not too hard to implement. Would need to add tests for this as well. It also mean that both contracts need to know the other contracts address so the deploy script would need updating. Give me your thoughts on this if you think it's too much. These challenges are meant to be noticeably harder than the SRE challenges.

**Review by escottalexander — CHANGES_REQUESTED** [review]
This is great, Fran. Your plot line blew me away! Left several small change suggestions and one big thought I had that is subject to discussion.

File: packages/foundry/contracts/MyToken.sol (hunk 11 above)
Comment (Franpastoragusti) [reply packages/foundry/contracts/MyToken.sol]: @escottalexander No problem to do it and agreed with the comments, but there is one thing, privacy. I will need to track who has vote what, in order to revert the action so who vote what will be expose

I was checking how to manage and I found some paths to follow but will complicate a lot the contract at this stage

Is fine for you to apply this functionality and explain by the narrative the complete transparent and no private result? in this way we can point this issue out and let the user think on it even we can create other challenges in the future to solve that

If you know a way to manage it that add some complexity but not too mutch for this level I'm open to sugestions

Comment (Franpastoragusti) [reply packages/foundry/contracts/MyToken.sol]: Same stuff with the amount, when the user votes the full balance is set as the vote value, when he transfers some founds should we directly remove all the vote weight or only the amount transfered at this stage of the challenge?

Is not a problem to apply it is that I don't want to set a too much complex challenge at the begining and I want to set well the requirements in each case
